File: backend/app/repositories/localities.py
from typing import Any

from motor.motor_asyncio import AsyncIOMotorDatabase

from app.repositories.base import MongoRepository


class LocalityRepository(MongoRepository):
    collection_name = "localities"

    def __init__(self, db: AsyncIOMotorDatabase):
        super().__init__(db[self.collection_name])

    async def find_by_slugs(self, slugs: list[str]) -> list[dict[str, Any]]:
        if not slugs:
            return []
        cursor = self.collection.find({"slug": {"$in": slugs}})
        return await cursor.to_list(length=len(slugs))

    async def top_for_city(self, city: str, limit: int = 12) -> list[dict[str, Any]]:
        cursor = self.collection.find({"city": city}).sort("scores.overall", -1).limit(limit)
        return await cursor.to_list(length=limit)

File: backend/src/repositories/localities.ts
import { Db } from "mongodb";
import { MongoRepository, serializeDoc } from "./base.js";

export class LocalityRepository extends MongoRepository {
  public static readonly COLLECTION_NAME = "localities";

  constructor(db: Db) {
    super(db.collection(LocalityRepository.COLLECTION_NAME));
  }

  async findBySlugs(slugs: string[]): Promise<any[]> {
    if (!slugs || slugs.length === 0) {
      return [];
    }
    const docs = await this.collection.find({ slug: { $in: slugs } } as any).toArray();
    return docs.map(serializeDoc);
  }

  async topForCity(city: string, limit: number = 12): Promise<any[]> {
    const docs = await this.collection
      .find({ city } as any)
      .sort({ "scores.overall": -1 })
      .limit(limit)
      .toArray();
    return docs.map(serializeDoc);
  }
}
